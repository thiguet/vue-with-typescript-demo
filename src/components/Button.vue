<template>
    <div class="wrapper-btn">
        <button :id="id" class="btn" :name="name" @click.prevent="onclick">
            <img class="icon" v-if="icon" :src="icon" :alt="name" />
            <span v-else>{{ label }} </span>
        </button>
    </div>
</template>

<script lang="ts">
import Vue from 'vue';

import { Prop, Component } from 'vue-property-decorator';

@Component
export default class Button extends Vue {
    @Prop({ required: true }) readonly id!: string;

    @Prop({ required: true }) readonly name!: string;

    @Prop({ required: true }) readonly onclick!: Function;

    @Prop() readonly label!: string;

    @Prop() readonly icon!: string;
}
</script>

<style scoped>
.wrapper-btn {
    display: flex;
    align-self: center;
    justify-content: center;
}

button {
    font-family: var(--font-family);
    font-size: var(--font-input-size);
    border: none;
    background-color: var(--second-bg);
    color: #fff;
    padding: 10px 15px;
    outline-offset: -4px;
    cursor: pointer;
}
</style>
