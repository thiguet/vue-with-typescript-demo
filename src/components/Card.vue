<template>
    <div class="container" @click="onclick">
        <img :src="icon" />
        <slot />
    </div>
</template>

<script lang="ts">
import Vue from 'vue';

import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';

@Component
export default class Card extends Vue {
    @Prop({ required: true }) readonly onclick!: Function;

    @Prop({ required: true }) readonly icon!: string;
}
</script>

<style>
.container {
    cursor: pointer;
    justify-content: center;
    padding: 0px;
    background-color: var(--ternary-bg);
    height: 70vh;
    width: 300px;
}

.container > img {
    max-width: 150px;
    max-height: 150px;
    align-self: center;
}

@media screen and (max-width: 600px) {
    .container {
        height: 32vh;
        padding: 10px;
        width: 85%;
        justify-content: center;
        align-self: center;
        margin-top: 20px;
    }

    .container > img {
        max-width: 120px;
        max-height: 120px;
    }
}
</style>
