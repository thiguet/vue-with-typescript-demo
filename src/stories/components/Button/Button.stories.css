@import url('../../app.css');

.btn {
    background-color: rgb(204, 52, 49) !important;
}

.icon {
    width: 100px;
    height: 100px;
}
