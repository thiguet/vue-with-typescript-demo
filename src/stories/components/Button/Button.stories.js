import Button from '@/components/Button.vue';
import './Button.stories.css';
import imageFile from './logo.gif';

export default {
    title: 'Example/Components/Button',
    component: Button,
};

const Template = (_args, { argTypes }) => ({
    props: Object.keys(argTypes),
    components: { Button },
    template: '<Button @onclick="onclick" v-bind="$props" />',
});

export const Default = Template.bind({});
Default.args = {
    id: 'some-id',
    name: 'some-name',
    label: 'Button',
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onclick: () => {},
};

export const WithIcon = Template.bind({});
WithIcon.args = {
    id: 'some-id2',
    name: 'some-name2',
    label: 'Button With Icon',
    icon: imageFile,
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onclick: () => {},
};
