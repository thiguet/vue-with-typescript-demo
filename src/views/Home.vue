<template>
    <div class="home">
        <div class="row">
            <Card
                :onclick="routeToProducts"
                icon="./assets/icons/supermarket-car.svg"
            >
            </Card>
            <Card
                :onclick="routeToReports"
                icon="./assets/icons/statistics.svg"
            >
            </Card>
        </div>
        <div class="row btn-row">
            <Button
                id="goback"
                class="goback"
                icon="./assets/icons/goback.svg"
                name="voltar"
                label="Voltar"
                :onclick="routeToLogin"
            />
        </div>
    </div>
</template>

<script lang="ts">
import { Vue, Component } from 'vue-property-decorator';

import Card from '@/components/Card.vue';
import Button from '@/components/Button.vue';

import { HomeView } from './models.d';

@Component({
    name: 'Home',
    components: {
        Card,
        Button,
    },
})
export default class Home extends Vue implements HomeView {
    public routeToProducts() {
        this.$router.push('/products');
    }

    public routeToReports() {
        this.$router.push('/reports');
    }

    public routeToLogin() {
        this.$router.push('/login');
    }
}
</script>

<style>
.home {
    flex-direction: column;
}
.btn-row {
    margin-top: 0px;
    align-self: center;
    max-width: 200px;
}

.row:nth-child(1) {
    min-height: 512px;
    overflow: auto;
}

.row:nth-child(2) {
    width: 100%;
    align-self: center;
}

.row {
    margin-top: 20px;
    justify-content: space-around;
}

@media screen and (max-width: 600px) {
    .row {
        flex-direction: column;
        justify-content: center;
    }
    .row .container:nth-child(2) {
        margin-top: 10px;
    }

    .row:nth-child(2) {
        margin-top: 40px;
    }
}

.goback {
    margin: auto;
}

.goback button {
    height: 60px;
    width: 150px;
    justify-content: center;
    display: flex;
}

.goback > button > .icon {
    align-self: center;
    width: 60px !important;
    filter: contrast(0) brightness(88);
}
</style>
