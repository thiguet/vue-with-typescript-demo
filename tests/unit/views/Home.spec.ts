import { shallowMount, mount, createLocalVue, Wrapper } from '@vue/test-utils';

import Vuex from 'vuex';

import Home from '@/views/Home.vue';

import Button from '@/components/Button.vue';
import Card from '@/components/Card.vue';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('Login', () => {
    const build = () => {
        const wrapper: Wrapper<Home> = shallowMount(Home, {
            localVue,
        });

        const mountedWrapper: Wrapper<Home> = mount(Home, {
            localVue,
        });

        return {
            wrapper,
            productsCard: () => mountedWrapper.findAllComponents(Card).at(0),
            reportsCard: () => mountedWrapper.findAllComponents(Card).at(1),
            goBackButton: () => mountedWrapper.findComponent(Button),
        };
    };

    beforeEach(() => {
    });

    it('renders component', () => {
        const { wrapper } = build();
        expect(wrapper).toMatchSnapshot();
    });

    it('renders main child components', () => {
        const { goBackButton, productsCard, reportsCard } = build();

        expect(productsCard().exists()).toBe(true);
        expect(reportsCard().exists()).toBe(true);
        expect(goBackButton().exists()).toBe(true);
    });
});
