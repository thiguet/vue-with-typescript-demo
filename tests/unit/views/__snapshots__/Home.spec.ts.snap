// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders component 1`] = `
<div class="home">
  <div class="row">
    <card-stub onclick="function () { [native code] }" icon="./assets/icons/supermarket-car.svg"></card-stub>
    <card-stub onclick="function () { [native code] }" icon="./assets/icons/statistics.svg"></card-stub>
  </div>
  <div class="row btn-row">
    <button-stub id="goback" name="voltar" onclick="function () { [native code] }" label="Voltar" icon="./assets/icons/goback.svg" class="goback"></button-stub>
  </div>
</div>
`;
